remix-solidity/src/compiler/compiler.ts

```diff
@@ -0,0 +1,392 @@
+'use strict'
+
+var solc = require('solc/wrapper')
+var solcABI = require('solc/abi')
+var webworkify = require('webworkify')
+import compilerInput from './compiler-input'
+var remixLib = require('remix-lib')
+var EventManager = remixLib.EventManager
+var txHelper = require('./txHelper')
+import { Source, SourceWithTarget, EVMVersion, Language, State, CompilationResult } from './types'
+
+/*
+  trigger compilationFinished, compilerLoaded, compilationStarted, compilationDuration
+*/
+export class Compiler {
+  event: any
+  state: State
+  handleImportCall: any
+
+  constructor (handleImportCall) {
+    this.event = new EventManager()
+    this.state = {
+      compileJSON: null,
+      worker: null,
+      currentVersion: null,
+      optimize: false,
+      evmVersion: null,
+      language: 'Solidity',
+      compilationStartTime: null,
+      target: null,
+      lastCompilationResult: {
+        data: null,
+        source: null
+      }
+    }
+    this.handleImportCall = handleImportCall
+
+    this.event.register('compilationFinished', (success, data, source) => {
+      if (success && this.state.compilationStartTime) {
+        this.event.trigger('compilationDuration', [(new Date().getTime()) - this.state.compilationStartTime])
+      }
+      this.state.compilationStartTime = null
+    })
+  
+    this.event.register('compilationStarted', () => {
+      this.state.compilationStartTime = new Date().getTime()
+    })
+  }
+
+  setOptimize (_optimize: boolean) {
```

Comment:
If you plan to format this is little you can easily do something like : 
```typescript
set <K extends keyof State>(key: K, value: State[K]) {
  this.state[key] = value
}
```
Or with an object input: 
```typescript
set (state: Partial<State>) {
  this.state = { ...this.state, ...state }
}
```

Reply:
done and also updated in IDE : https://github.com/ethereum/remix-ide/pull/2511